TEST: 1.19.x + blas variants

h-vetinari opened the pull request:
Based on #195; doing a similar investigation as in conda-forge/scipy-feedstock#130.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
This looks quite a bit better than on the scipy-side. Still it did catch some errors that might not show up anywhere else in CI.

I mentioned in https://github.com/conda/conda-build/issues/3947 that IMO, foundational packages like numpy/scipy should be tested across all available blas variants. This would - effectively - provide some further CI on the various blas builds, but also make sure that no broken variants are shipped.

Note: The cells of the matrix have the same behaviour for all python versions (cpython 3.6-3.8, pypy 3.6) unless otherwise noted - hence the simplified matrix

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86| ✔️| ✔️ | ✔️ (cpython) / ❌ (pypy) | ✔️ (cpython) / ❌ (pypy) |
| linux / aarch | ➖ | ➖ | ✔️  | ❌ (non-uniformly) |
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ |
| osx | ✔️ | ✔️ | ✔️ | ✔️ |
| win | ✔️ | ✔️ | ✔️ | ✔️ |

There were 4 additional failures not shown in the matrix above:
* the failure for osx+mkl on python 3.8 is a spurious http error
* aarch+netlib+pypy runs into 60min drone timeout at ~54% of the test suite
* ppc64le+netlib had two build script errors on travis, but I don't have rights to restart the job


<details>
<summary>linux + x86 + netlib/openblas + pypy: segfault, test suite aborted</summary>

```
linalg/tests/test_linalg.py::TestCond::test_sq_cases PASSED              [ 62%]
linalg/tests/test_linalg.py::TestCond::test_empty_sq_cases PASSED        [ 62%]
Fatal Python error: Segmentation fault

Stack (most recent call first, approximate line numbers):
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 81 in check [jitted]
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 337 in check_cases [jitted]
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 337 in check_cases
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 392 in test_generalized_sq_cases
  File "[...snip...]/site-packages/_pytest/python.py", line 175 in pytest_pyfunc_call
  File "[...snip...]/site-packages/pluggy/callers.py", line 157 in _multicall [jit inlined]
  File "[...snip...]/site-packages/pluggy/manager.py", line 84 in <lambda> [jit inlined]
  File "[...snip...]/site-packages/pluggy/manager.py", line 90 in _hookexec [jitted]
  File "[...snip...]/site-packages/pluggy/hooks.py", line 272 in __call__ [jitted]
  File "[...snip...]/site-packages/pluggy/hooks.py", line 272 in __call__
  File "[...snip...]/site-packages/_pytest/python.py", line 1475 in runtest
  File "[...snip...]/site-packages/_pytest/runner.py", line 126 in pytest_runtest_call
  File "[...snip...]/site-packages/pluggy/callers.py", line 157 in _multicall [jit inlined]
  File "[...snip...]/site-packages/pluggy/manager.py", line 84 in <lambda> [jit inlined]
Tests failed for numpy-1.19.0-py36h04dfcbd_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for numpy-1.19.0-py36h04dfcbd_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING conda_build.build:tests_failed(2751): Tests failed for numpy-1.19.0-py36h04dfcbd_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: numpy-1.19.0-py36h04dfcbd_0.tar.bz2
```

</details>

<details>
<summary>linux + aarch + openblas + cpython 3.6: 1 test failure (TestCholesky.test_basic_property)</summary>

```
=================================== FAILURES ===================================
_______________________ TestCholesky.test_basic_property _______________________

self = <numpy.linalg.tests.test_linalg.TestCholesky object at 0x706e90d77b38>

    def test_basic_property(self):
        # Check A = L L^H
        shapes = [(1, 1), (2, 2), (3, 3), (50, 50), (3, 10, 10)]
        dtypes = (np.float32, np.float64, np.complex64, np.complex128)
    
        for shape, dtype in itertools.product(shapes, dtypes):
            np.random.seed(1)
            a = np.random.randn(*shape)
            if np.issubdtype(dtype, np.complexfloating):
                a = a + 1j*np.random.randn(*shape)
    
            t = list(range(len(shape)))
            t[-2:] = -1, -2
    
            a = np.matmul(a.transpose(t).conj(), a)
            a = np.asarray(a, dtype=dtype)
    
            c = np.linalg.cholesky(a)
    
            b = np.matmul(c, c.transpose(t).conj())
            assert_allclose(b, a,
                            err_msg="{} {}\n{}\n{}".format(shape, dtype, a, c),
>                           atol=500 * a.shape[0] * np.finfo(dtype).eps)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=5.55112e-12
E           (50, 50) <class 'numpy.complex128'>
E           [[ 86.08558859 +0.j           3.44218047+15.88127244j
E              -0.45363563 -0.44661413j ...   2.64035992 -9.58169853j
E              -3.23064882 +6.99917231j   3.72228944 -3.50497908j]
E            [  3.44218047-15.88127244j  97.31130762 +0.j
E               4.95610003 +3.51207641j ... -24.44048428 -0.53073368j
E             -21.33047678 +4.24795683j  -2.7632562  +5.87612999j]
E            [ -0.45363563 +0.44661413j   4.95610003 -3.51207641j
E             108.58323952 +0.j         ...   6.24573557 -3.07029841j
E               7.36481721 -5.16506877j  -1.1160803  -0.23457131j]
E            ...
E            [  2.64035992 +9.58169853j -24.44048428 +0.53073368j
E               6.24573557 +3.07029841j ...  97.33841401 +0.j
E              13.64351655 +0.62184078j   3.54722947 +1.06034335j]
E            [ -3.23064882 -6.99917231j -21.33047678 -4.24795683j
E               7.36481721 +5.16506877j ...  13.64351655 -0.62184078j
E              99.49995802 +0.j         -15.73440912 -2.65289272j]
E            [  3.72228944 +3.50497908j  -2.7632562  -5.87612999j
E              -1.1160803  +0.23457131j ...   3.54722947 -1.06034335j
E             -15.73440912 +2.65289272j  89.70910493 +0.j        ]]
E           [[ 9.27823198+0.j          0.        +0.j          0.        +0.j
E             ...  0.        +0.j          0.        +0.j
E              0.        +0.j        ]
E            [ 0.3709953 -1.71167012j  9.70792746+0.j          0.        +0.j
E             ...  0.        +0.j          0.        +0.j
E              0.        +0.j        ]
E            [-0.04889246+0.04813569j  0.52087651-0.35499305j 10.40101916+0.j
E             ...  0.        +0.j          0.        +0.j
E              0.        +0.j        ]
E            ...
E            [ 0.28457576+1.03270737j -2.3463717 -0.03497098j  0.71336226+0.38319797j
E             ...  2.96242819+0.j          0.        +0.j
E              0.        +0.j        ]
E            [-0.3481966 -0.75436488j -2.31692309-0.3473546j   0.81411515+0.58790833j
E             ...  0.99318458+0.51943463j  1.12603697+0.j
E              0.        +0.j        ]
E            [ 0.40118521+0.37776368j -0.23336466-0.69046409j -0.11904651+0.06872805j
E             ...  0.0339795 -0.51402601j  0.96743148-0.46502072j
E              1.24939718+0.j        ]]
E           Mismatched elements: 436 / 2500 (17.4%)
E           Max absolute difference: 94.66308979
E           Max relative difference: 16.00607492
E            x: array([[ 86.085589 +0.j      ,   3.44218 +15.881272j,
E                    -0.453636 -0.446614j, ...,   2.64036  -9.581699j,
E                    -3.230649 +6.999172j,   3.722289 -3.504979j],...
E            y: array([[ 86.085589 +0.j      ,   3.44218 +15.881272j,
E                    -0.453636 -0.446614j, ...,   2.64036  -9.581699j,
E                    -3.230649 +6.999172j,   3.722289 -3.504979j],...

../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py:1736: AssertionError
=============================== warnings summary ===============================
```
</details>

<details>
<summary>linux + aarch + openblas + cpython 3.7: 4 test failures</summary>

```
=========================== short test summary info ============================
FAILED core/tests/test_multiarray.py::TestDot::test_dot_3args - AssertionError: 
FAILED linalg/tests/test_linalg.py::TestCond::test_sq_cases - AssertionError:...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt3]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - numpy...
= 4 failed, 10394 passed, 568 skipped, 26 deselected, 19 xfailed, 5 xpassed, 233 warnings in 424.23s (0:07:04) =
```
</details>

<details>
<summary>linux + aarch + openblas + cpython 3.8: 3 test failures</summary>

```
=========================== short test summary info ============================
FAILED core/tests/test_multiarray.py::TestDot::test_dot_3args - AssertionError: 
FAILED linalg/tests/test_linalg.py::TestCond::test_sq_cases - AssertionError:...
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - numpy...
= 3 failed, 10398 passed, 567 skipped, 26 deselected, 19 xfailed, 5 xpassed, 233 warnings in 401.93s (0:06:41) =
```
</details>

h-vetinari commented:
Maybe worth pinging @conda-forge/help-pypy on this. The segfaults on linux+pypy for numpy 1.19 affect both this PR and #195.

mattip commented:
What version of OpenBLAS are you testing with? 0.3.10 was released recently and fixes an interaction that would cause a segfault on PyPy on azure machines.

h-vetinari commented:
@mattip 
It was running with `libopenblas 0.3.9` before, but even after bumping to `0.3.10` (see https://github.com/conda-forge/blas-feedstock/pull/45), the segfault with openblas persists. Interestingly (and I have no good reason why), the segfault is now gone for the netlib variant (perhaps it's not fully deterministic).

There's more stuff that I don't understand, actually. Bumping blis from 0.6.1 to 0.7.0 has introduced a segfault on cpython 3.6, 12 test failures on 3.7, but runs through fine on 3.8...? 🤔

h-vetinari commented:
After bumping blis to 0.7.0 and openblas to 0.3.10 (see https://github.com/conda-forge/blas-feedstock/pull/45), the picture changed a bit, in some ways unexpectedly.

In short:
* the linux+x86+pypy segfault remains, but seems non-deterministic (in a run before it disappeared only from the netlib build, not the openblas one, now it _did_ appeared for the first time in the mkl build, while disappearing for the others)
* linux+ppc64le+openblas failures remained
* both pypy builds on aarch time out
* win+blis is now failing
* TestCholesky.test_basic_property seems to be the most common failure across variants
* the failure of win+mkl+py37 is likely spurious

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86| ✔️| ✔️ (cpython) / ❌ (pypy) | ✔️ | ✔️ |
| linux / aarch | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy)  | ✔️ (cpython) / ❌ (pypy) |
| linux / ppc64le | ➖ | ➖ | ✔️ | ❌ (non-uniformly) |
| osx | ✔️ | ✔️ | ✔️ | ✔️ |
| win | ❌ | ✔️ | ✔️ | ✔️ |


<details>
<summary>linux + x86 + mkl + pypy: segfault, test suite aborted</summary>

```
linalg/tests/test_linalg.py::TestCond::test_sq_cases PASSED              [ 62%]
linalg/tests/test_linalg.py::TestCond::test_empty_sq_cases PASSED        [ 62%]
Fatal Python error: Segmentation fault

Stack (most recent call first, approximate line numbers):
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 81 in check [jitted]
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 337 in check_cases [jitted]
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 337 in check_cases
  File "[...snip...]/lib/python3.6/site-packages/numpy/linalg/tests/test_linalg.py", line 392 in test_generalized_sq_cases
  File "[...snip...]/site-packages/_pytest/python.py", line 175 in pytest_pyfunc_call
  File "[...snip...]/site-packages/pluggy/callers.py", line 157 in _multicall [jit inlined]
  File "[...snip...]/site-packages/pluggy/manager.py", line 84 in <lambda> [jit inlined]
  File "[...snip...]/site-packages/pluggy/manager.py", line 90 in _hookexec [jitted]
  File "[...snip...]/site-packages/pluggy/hooks.py", line 272 in __call__ [jitted]
  File "[...snip...]/site-packages/pluggy/hooks.py", line 272 in __call__
  File "[...snip...]/site-packages/_pytest/python.py", line 1475 in runtest
  File "[...snip...]/site-packages/_pytest/runner.py", line 126 in pytest_runtest_call
  File "[...snip...]/site-packages/pluggy/callers.py", line 157 in _multicall [jit inlined]
  File "[...snip...]/site-packages/pluggy/manager.py", line 84 in <lambda> [jit inlined]
Tests failed for numpy-1.19.0-py36h04dfcbd_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for numpy-1.19.0-py36h04dfcbd_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING conda_build.build:tests_failed(2751): Tests failed for numpy-1.19.0-py36h04dfcbd_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: numpy-1.19.0-py36h04dfcbd_0.tar.bz2
```

</details>

<details>
<summary>linux + ppc64le + openblas + cpython: 3 test failures</summary>

```
=========================== short test summary info ============================
FAILED linalg/tests/test_linalg.py::TestCond::test_sq_cases - AssertionError:...
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - numpy...
FAILED linalg/tests/test_linalg.py::TestTensorinv::test_tensorinv_result - As...
= 3 failed, 10395 passed, 568 skipped, 26 deselected, 19 xfailed, 5 xpassed, 233 warnings in 507.77s (0:08:27) =
```
</details>

<details>
<summary>linux + ppc64le + openblas + pypy: 4 test failures</summary>

```
=========================== short test summary info ============================
FAILED core/tests/test_multiarray.py::TestDot::test_dot_3args - AssertionError: 
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestCond::test_sq_cases - AssertionError:...
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - numpy...
= 4 failed, 10337 passed, 612 skipped, 26 deselected, 21 xfailed, 5 xpassed, 233 warnings in 993.28s (0:16:33) =
```
</details>

<details>
<summary>windows + blis + cpython: 11-12 failed tests</summary>

This one seems to be a bit fickle as well - in the previous run, it appeared for py36 & py37 but not py38, not it appeared for py37 & py38 but not py36.

```
=========================== short test summary info ===========================
FAILED core/tests/test_multiarray.py::TestMatmul::test_matmul_object - Assert...
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestPinv::test_nonsq_cases - AssertionErr...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_sq_cases - Ass...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_nonsq_cases - ...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 12 failed, 10401 passed, 493 skipped, 19 xfailed, 1 xpassed, 228 warnings in 545.60s (0:09:05) =
```
</details>

mattip commented:
Thanks for the summary. Numpy regularly builds on travisCI machines with PyPy, both in the numpy/numpy CI and on the MacPython/num-wheels CI. I wonder what is different on the Azure builds?

mattip commented:
In the [build log for pypy](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=179206&view=logs&j=696704cc-6fef-57a3-ea36-f27779b8cd5e&t=06421391-4b55-523d-a804-5e1a5bfc1908&l=9872) I am seeing [`libcblas:           3.8.0-16_mkl`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=179206&view=logs&j=696704cc-6fef-57a3-ea36-f27779b8cd5e&t=06421391-4b55-523d-a804-5e1a5bfc1908&l=2717). Am I looking at the wrong logs? I expected to see 3.10.

h-vetinari commented:
> @mattip: In the build log for pypy I am seeing `libcblas: 3.8.0-16_mkl`. Am I looking at the wrong logs? I expected to see 3.10.

You're looking at the right logs, but I think you're mixing up the reference-lapack [version numbers](https://github.com/Reference-LAPACK/lapack/releases) with openblas (note: 3.9.0 isn't in conda-forge yet because MKL isn't compatible yet). `libblas` / `libcblas` / `liblapack` / `liblapacke` get the lapack version number, and a build string suffix indicating which blas flavour (`blis` / `mkl` / `netlib` = reference-lapack / `openblas`) - you were looking at a run with mkl.

As I noted in the last comment, the segfault on pypy seems non-deterministic. I appeared for openblas 2/3 run, for netlib 1/3 times and for mkl 1/3 times. One of the runs has been deleted by azure, but [here's](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=178823&view=results) one where the segfault occurred with openblas. I haven't investigated which test follows `linalg/tests/test_linalg.py::TestCond::test_empty_sq_cases PASSED        [ 62%]`, but it always seems to happen there.

mattip commented:
Ahh, right sorry. But in [the `linux_blas_implopenblasnumpy1.18python3.6.____73_pypytarget_platformlinux-64.yaml` log](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=178823&view=logs&j=bbe49da5-7bd7-508d-ebc8-759d9757842e&t=12793a5b-c868-5fad-4a77-9d9537fc1e96&l=2723), I see `libopenblas:        0.3.9-h5ec1e0e_0`. Shouldn't that be 0.3.10?

In NumPy we use add a check like this (from [`tools/openblas.py`](https://github.com/numpy/numpy/blob/v1.19.0/tools/openblas_support.py#L280):

```
def test_version(expected_version, ilp64=get_ilp64()):
    """
    Assert that expected OpenBLAS version is
    actually available via NumPy
    """
    import numpy
    import ctypes

    dll = ctypes.CDLL(numpy.core._multiarray_umath.__file__)
    if ilp64 == "64_":
        get_config = dll.openblas_get_config64_
    else:
        get_config = dll.openblas_get_config
    get_config.restype=ctypes.c_char_p
    res = get_config()
    print('OpenBLAS get_config returned', str(res))
    if not expected_version:
        expected_version = OPENBLAS_V
    check_str = b'OpenBLAS %s' % expected_version.encode()
    print(check_str)
    assert check_str in res, '%s not found in %s' %(expected_version, res)
    if ilp64:
        assert b"USE64BITINT" in res
    else:
        assert b"USE64BITINT" not in res
```

h-vetinari commented:
> @mattip: Shouldn't that be 0.3.10?

You're right, sorry. The run I linked was before conda-forge/blas-feedstock#45 was merged. The first run after that _had_ the segfault, but got deleted by azure since I restarted before everything was through. Let's try that again.

mattip commented:
this is a tough PR. The starts must align just perfectly for everything to pass in the same run.

h-vetinari commented:
> this is a tough PR. The starts must align just perfectly for everything to pass in the same run.

It seems to me the best answer would be to figure out what causes the segfault and fix it 😉 
So far, the pypy build has always failed for at least one of the blas variants.

The blis+win builds are also failing consistently, as are the linux+ppc builds. So we'll probably need more than lucky stars to pull this into the green 😅

mattip commented:
I wonder if there is a path to release a 1.19 NumPy for at least the most common platforms.

mattip commented:
I see that OpenBLAS is being built with `DYNAMIC_ARCH`. Is a base `TARGET being set as well? See for instance MacPython/openblas-libs#34

<details>
In [this comment](https://github.com/xianyi/OpenBLAS/issues/2388#issuecomment-582267407), and elsewhere, they recommends using `TARGET` as a minimal platform when `DYNAMIC_ARCH` is used which will be the fallback when no better dynamic loop can be found. The default is the build platform, which may be problematic if the build is done on a more advanced machine than the test.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict.
Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.

isuruf commented:
@conda-forge-admin, rerender

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
Update:
* the linux+x86+pypy segfault: did not reappear
* linux+ppc64le+openblas failures are gone
* both pypy builds on aarch _still_ time out
* win+blis is still failing, but with a different error
* the failure of win+openblas+py36 is the same error as win+blis

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86| ✔️| ✔️ | ✔️ | ✔️ |
| linux / aarch | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy)  | ✔️ (cpython) / ❌ (pypy) |
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ |
| osx | ✔️ | ✔️ | ✔️ | ✔️ |
| win | ✔️ (py36) / ❌ (py37,py38) | ✔️ | ✔️ | ✔️ (py37,py38) / ❌ (py36) |


<details>
<summary>win + (blis,py37)/(blis,py38)/(openblas,py36): error about nonexistent pipe</summary>

```
core/tests/test_multiarray.py::TestIO::test_binary PASSED                [ 17%]
Traceback (most recent call last):
  File "C:\Miniconda\Scripts\conda-build-script.py", line 10, in <module>
    sys.exit(main())
  File "C:\Miniconda\lib\site-packages\conda_build\cli\main_build.py", line 474, in main
    execute(sys.argv[1:])
  File "C:\Miniconda\lib\site-packages\conda_build\cli\main_build.py", line 465, in execute
    verify=args.verify, variants=args.variants)
  File "C:\Miniconda\lib\site-packages\conda_build\api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 2870, in build_tree
    test(pkg, config=metadata.config.copy(), stats=stats)
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 2712, in test
    utils.check_call_env(cmd, env=env, cwd=metadata.config.test_dir, stats=test_stats, rewrite_stdout_env=rewrite_env)
  File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 404, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 379, in _func_defaulting_env_to_os_environ
    proc = PopenWrapper(_args, **kwargs)
  File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 264, in __init__
    self.out, self.err = self._execute(*args, **kwargs)
  File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 301, in _execute
    for child in parent.children(recursive=True):
  File "C:\Miniconda\lib\site-packages\psutil\__init__.py", line 289, in wrapper
    return fun(self, *args, **kwargs)
  File "C:\Miniconda\lib\site-packages\psutil\__init__.py", line 919, in children
    ppid_map = _ppid_map()
OSError: [WinError 1455] The paging file is too small for this operation to complete
Traceback (most recent call last):
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\main.py", line 191, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\main.py", line 247, in _main
    config.hook.pytest_runtestloop(session=session)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\main.py", line 272, in pytest_runtestloop
    item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\runner.py", line 84, in pytest_runtest_protocol
    item.ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\terminal.py", line 439, in pytest_runtest_logstart
    self.write_ensure_prefix(line, "")
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\terminal.py", line 350, in write_ensure_prefix
    self._tw.line()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 273, in line
    self.write('\n')
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 256, in write
    write_out(self._file, markupmsg)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 406, in write_out
    fil.write(msg)
OSError: [Errno 22] Invalid argument

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\main.py", line 213, in wrap_session
    config.notify_exception(excinfo, config.option)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\config\__init__.py", line 866, in notify_exception
    res = self.hook.pytest_internalerror(excrepr=excrepr, excinfo=excinfo)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\terminal.py", line 407, in pytest_internalerror
    self.write_line("INTERNALERROR> " + line)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\terminal.py", line 368, in write_line
    self.ensure_newline()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\terminal.py", line 359, in ensure_newline
    self._tw.line()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 273, in line
    self.write('\n')
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 256, in write
    write_out(self._file, markupmsg)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 406, in write_out
    fil.write(msg)
OSError: [Errno 22] Invalid argument

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\bld\numpy_1594940373640\_test_env\Scripts\pytest-script.py", line 10, in <module>
    sys.exit(main())
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\config\__init__.py", line 125, in main
    config=config
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\main.py", line 240, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\main.py", line 229, in wrap_session
    session=session, exitstatus=session.exitstatus
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\pluggy\callers.py", line 203, in _multicall
    gen.send(outcome)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\_pytest\terminal.py", line 718, in pytest_sessionfinish
    self._tw.line("")
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 273, in line
    self.write('\n')
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 256, in write
    write_out(self._file, markupmsg)
  File "D:\bld\numpy_1594940373640\_test_env\lib\site-packages\py\_io\terminalwriter.py", line 406, in write_out
    fil.write(msg)
OSError: [Errno 22] Invalid argument
##[error]Cmd.exe exited with code '1'.
The process tried to write to a nonexistent pipe.
The process tried to write to a nonexistent pipe.
```
</details>

h-vetinari commented:
Update (after bumping to 1.19.1)
* the linux+x86+pypy segfault reappeared for an mkl build
* pypy+aarch now only times out for the netlib variant
* win+blis is not hard-aborting with a broken pipe anymore, but back to the 11-13 test failures

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86| ✔️| ✔️ (cpython) / ❌ (pypy) | ✔️ | ✔️ |
| linux / aarch | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy)  | ✔️ |
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ |
| osx | ✔️ | ✔️ | ✔️ | ✔️ |
| win | ❌ | ✔️ | ✔️ | ✔️ |


<details>
<summary>win + blis: 11-13 errors</summary>

```
=========================== short test summary info ===========================
FAILED core/tests/test_multiarray.py::TestMatmul::test_matmul_object - Assert...
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestPinv::test_nonsq_cases - AssertionErr...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_sq_cases - Ass...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_nonsq_cases - ...
FAILED linalg/tests/test_linalg.py::TestPinvHermitian::test_generalized_herm_cases
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 13 failed, 10470 passed, 493 skipped, 19 xfailed, 1 xpassed, 230 warnings in 421.72s (0:07:01) =
```

</details>

h-vetinari commented:
@conda-forge/numpy 
Not sure what changed since last time (I just rerendered), but now, it seems every single build is failing with:
```
=================================== FAILURES ===================================
_____________________________ test_api_importable ______________________________

    def test_api_importable():
        """
        Check that all submodules listed higher up in this file can be imported
    
        Note that if a PRIVATE_BUT_PRESENT_MODULES entry goes missing, it may
        simply need to be removed from the list (deprecation may or may not be
        needed - apply common sense).
        """
        def check_importable(module_name):
            try:
                importlib.import_module(module_name)
            except (ImportError, AttributeError):
                return False
    
            return True
    
        module_names = []
        for module_name in PUBLIC_MODULES:
            if not check_importable(module_name):
                module_names.append(module_name)
----------------------------- Captured stdout call -----------------------------
Could not locate executable icc
Could not locate executable ecc
=============================== warnings summary ===============================
```

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict.
Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
Update:
* the flaky linux+x86+pypy segfault reappeared for openblas, disappeared for mkl
* pypy+aarch times out again for both netlib & openblas
* pypy+ppc64le+openblas times out now
* win+blis is still failing, but number of test errors was reduced from 11-13 to 8.
* likely spurious test abort under win+mkl+cpython 3.7 with `OSError: [WinError 1455] The paging file is too small for this operation to complete`
* looks like osx+netlib+pypy is hanging on a test and will time out

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86| ✔️| ✔️ | ✔️ | ✔️ (cpython) / ❌ (pypy) |
| linux / aarch | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy)  | ✔️ (cpython) / ❌ (pypy) |
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ (cpython) / ❌ (pypy) |
| osx | ✔️ | ✔️ | ✔️ (cpython) / ❌ (pypy) | ✔️ |
| win | ❌ | ✔️ | ✔️ | ✔️ |

[Build logs](https://github.com/conda-forge/numpy-feedstock/runs/1078558735)

<details>
<summary>win + blis: 8 errors</summary>

```
=========================== short test summary info ===========================
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 8 failed, 10475 passed, 493 skipped, 19 xfailed, 1 xpassed, 227 warnings in 574.89s (0:09:34) =
```

</details>

h-vetinari commented:
Update after upgrading to 1.19.2:
* the flaky linux+x86+pypy segfault remained for openblas
* pypy+aarch times out netlib, not for openblas anymore
* pypy+ppc64le+openblas ran through again
* win+blis is now **passing** except a likely spurious `OSError: [WinError 1455] The paging file is too small for this operation to complete` under cpython 3.6

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86| ✔️| ✔️ | ✔️ | ✔️ (cpython) / ❌ (pypy) |
| linux / aarch | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy) | ✔️ |
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ |
| osx | ✔️ | ✔️ | ✔️ | ✔️ |
| win | ✔️ | ✔️ | ✔️ | ✔️ |

[Build logs](https://github.com/conda-forge/numpy-feedstock/runs/1104161193)

h-vetinari commented:
Update (after bumping to 1.19.3)
* the pypy+linux+x86 segfault disappeared again
* pypy+linux+ppc64le+openblas has gained a segfault
* pypy+linux+aarch timout moved from netlib to openblas
* win+blis is back to some of the old test failures, but only on python 3.6
* non-relevant azure timeouts on osx+x86+blis

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86 | ✔️| ✔️ | ✔️ | ✔️ |
| linux / aarch | ➖ | ➖ | ✔️ | ✔️ (cpython) / ❌ (pypy) |
| linux / ppc64le | ➖ | ➖ | ✔️ | ✔️ (cpython) / ❌ (pypy) |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ |
| win | ✔️ (py3.7-3.9) / ❌ (py3.6) | ✔️ | ✔️ | ✔️ |

[Build logs](https://github.com/conda-forge/numpy-feedstock/pull/196/checks?check_run_id=1326482136)

<details>
<summary>linux + ppc64le + openblas + pypy: segfault</summary>

```
lib/tests/test_format.py::test_bad_header PASSED                         [ 51%]
lib/tests/test_format.py::test_large_file_support PASSED                 [ 51%]
/home/conda/feedstock_root/build_artifacts/numpy_1603975825295/test_tmp/run_test.sh: line 9:  2173 Killed                  pytest --verbose --pyargs numpy -k "not (_not_a_real_test or test_einsum_sums_cfloat64 or test_loss_of_precision or test_large_zip or test_may_share_memory_easy_fuzz or test_may_share_memory_harder_fuzz or test_unary_ufunc_call_fuzz or test_count_nonzero_all or test_diophantine_fuzz or test_generalized_sq_cases or test_may_share_memory_harder_fuzz or test_large_zip)" --durations=0
Tests failed for numpy-1.19.3-py36h8a64ab2_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for numpy-1.19.3-py36h8a64ab2_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING conda_build.build:tests_failed(2895): Tests failed for numpy-1.19.3-py36h8a64ab2_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: numpy-1.19.3-py36h8a64ab2_0.tar.bz2
The command "if [[ ${PLATFORM} =~ .*linux.* ]]; then ./.scripts/run_docker_build.sh; fi" exited with 1.
```

</details>

<details>
<summary>win + blis + cpython 3.6: 8 errors</summary>

```
=========================== short test summary info ===========================
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 8 failed, 10476 passed, 494 skipped, 19 xfailed, 1 xpassed, 227 warnings in 614.93s (0:10:14) =
```

</details>

h-vetinari commented:
Update (after bumping to 1.19.4)
* pypy+linux+x86 segfault did *not* reappear 🥳 
* pypy+linux+ppc64le segfault moved from openblas to netlib
* pypy+linux+aarch timout disappeared
* win+blis test failures, now only on python 3.7 (instead of 3.6 for 1.19.3) - seems flaky as well?
* 3 errors `The process tried to write to a nonexistent pipe.` on windows - these seem to be annoyingly persistent (also in #210)

| variant| blis | mkl | netlib | openblas |
|-----|------|----|----|----|
| linux / x86 | ✔️| ✔️ | ✔️ | ✔️ |
| linux / aarch | ➖ | ➖ | ✔️ | ✔️ |
| linux / ppc64le | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy) | ✔️ |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ |
| win | ✔️ (3.6,3.8,3.9) / ❌ (3.7) | ✔️ | ✔️ | ✔️ |

[Build logs](https://github.com/conda-forge/numpy-feedstock/pull/196/checks?check_run_id=1326482136)

<details>
<summary>linux + ppc64le + netlib + pypy: segfault</summary>

```
lib/tests/test_format.py::test_bad_header PASSED                         [ 51%]
lib/tests/test_format.py::test_large_file_support PASSED                 [ 51%]
/home/conda/feedstock_root/build_artifacts/numpy_1604658271803/test_tmp/run_test.sh: line 9:  2104 Killed                  pytest --verbose --pyargs numpy -k "not (_not_a_real_test or test_einsum_sums_cfloat64 or test_loss_of_precision or test_large_zip or test_may_share_memory_easy_fuzz or test_may_share_memory_harder_fuzz or test_unary_ufunc_call_fuzz or test_count_nonzero_all or test_diophantine_fuzz or test_generalized_sq_cases or test_may_share_memory_harder_fuzz or test_large_zip)" --durations=0
Tests failed for numpy-1.19.4-py36he8f70c9_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for numpy-1.19.4-py36he8f70c9_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING conda_build.build:tests_failed(2895): Tests failed for numpy-1.19.4-py36he8f70c9_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: numpy-1.19.4-py36he8f70c9_0.tar.bz2
The command "if [[ ${PLATFORM} =~ .*linux.* ]]; then ./.scripts/run_docker_build.sh; fi" exited with 1.
```

</details>

<details>
<summary>win + blis + cpython 3.7: 8 errors</summary>

```
=========================== short test summary info ===========================
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 8 failed, 10476 passed, 494 skipped, 19 xfailed, 1 xpassed, 227 warnings in 454.64s (0:07:34) =
```

</details>

h-vetinari commented:
### Going from 1.19.4 to 1.19.5 

From 2 failures out of 80 combinations, there are now 5 failures out of 80.

### Details

The builds are now using `blis 0.8.0` instead of `0.7.0`.

| variant | status 1.19.4 |  status 1.19.5 |
|--------|-------|----|
| linux + ppc64le + pypy + netlib | segfault | segfault remains |
| win + blis | some test errors under py3.7 | 11-12 failures across all python versions |

---------------

| variant| blis | mkl | netlib | openblas | sum* | 
|-----|------|----|----|----|----|
| linux / x86 | ✔️| ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ✔️ (cpython) / ❌ (pypy) | ✔️ | 1F |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ | - |
| win | ❌ | ✔️ | ✔️ | ✔️ | 4F |
| sum* | 4F | - | 1F | - | 5F |

`*` sum of Failures (out of a total of 80 CI combinations being tested)

[Build logs](https://github.com/conda-forge/numpy-feedstock/pull/196/checks?check_run_id=1665716110)

<details>
<summary>linux + ppc64le + netlib + pypy: segfault</summary>

```
lib/tests/test_format.py::test_bad_header PASSED                         [ 51%]
lib/tests/test_format.py::test_large_file_support PASSED                 [ 51%]
/home/conda/feedstock_root/build_artifacts/numpy_1610062633896/test_tmp/run_test.sh: line 9:  2175 Killed                  pytest --verbose --pyargs numpy -k "not (_not_a_real_test or test_einsum_sums_cfloat64 or test_loss_of_precision or test_large_zip or test_may_share_memory_easy_fuzz or test_may_share_memory_harder_fuzz or test_unary_ufunc_call_fuzz or test_count_nonzero_all or test_diophantine_fuzz or test_generalized_sq_cases or test_may_share_memory_harder_fuzz or test_large_zip)" --durations=0
Tests failed for numpy-1.19.5-py36h0f9a637_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for numpy-1.19.5-py36h0f9a637_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING conda_build.build:tests_failed(2895): Tests failed for numpy-1.19.5-py36h0f9a637_0.tar.bz2 - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: numpy-1.19.5-py36h0f9a637_0.tar.bz2
The command "if [[ ${PLATFORM} =~ .*linux.* ]]; then ./.scripts/run_docker_build.sh; fi" exited with 1.
```

</details>

<details>
<summary>win + blis: 11-12 errors</summary>

```
=========================== short test summary info ===========================
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestPinv::test_nonsq_cases - AssertionErr...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_sq_cases - Ass...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_nonsq_cases - ...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 11 failed, 10487 passed, 494 skipped, 19 xfailed, 1 xpassed, 229 warnings in 486.34s (0:08:06) =
```

</details>

h-vetinari commented:
### Update with pypy37, as well as...
| lib | before | after |
|--------|-------|----|
| `numpy` | `1.19.5` | `1.19.5` |
| `libblas` | `3.9.0-7` |  `3.9.0-7` |
| `blis` | `0.8.0-1` |  `0.8.0-1` |
| `openblas` | `0.3.12-pthreads-1`| `0.3.12-pthreads-1` | 
| `mkl` | `2020.4-304` | `2020.4-304` | 
| `netlib` | `3.9.0-3` | `3.9.0-3` | 

From 6 failures out of 80 combinations, there are now 4 failures out of 92. [travis bombed completely, not counting those 12 runs yet, waiting for rerun]

#### The good news:
* pypy37 doing well
* first (single) passing run on win + blis in a while 🥳 

### Details

| variant | before | after |
|--------|-------|----|
| linux + aarch + pypy3.6 + openblas  | one test failure (probably flaky as it's "only" from hypothesis) | same failure remains |
| linux + ppc64le + pypy + netlib | segfault | TBD |
| win + blis | 11-12 failures across all python versions | first passed test-suite with py38 only, for reasons unclear; 11-13 failures otherwise |
| win | ... | 3 failures due to `The process tried to write to a nonexistent pipe.` across blis/mkl/openblas |

---------------

| variant| blis | mkl | netlib | openblas | sum* | 
|-----|------|----|----|----|----|
| linux / x86 | ✔️| ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ✔️ (cpython, pypy37) / ❌ (pypy36) | ✔️ | 1F | 
| linux / ppc64le | ➖ | ➖ | ... | ... |  |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | - |
| osx / arm | ➖ | ➖ | ✔️ | ✔️ | - |
| win | ✔️ (py38) / ❌ (py36, py37, py39) | ✔️ | ✔️ | ✔️ | 3F |
| sum* | 3F | - | 1F | - | 4F |

`*` sum of Failures (out of a total of 92 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=269697&view=results)
[Drone](https://cloud.drone.io/conda-forge/numpy-feedstock/170/11/2)
[Travis](https://travis-ci.com/github/conda-forge/numpy-feedstock/builds/214703600)

<details>
<summary>linux + aarch + netlib + pypy: 1 test failure</summary>

```
=================================== FAILURES ===================================
_________________________ TestClip.test_clip_property __________________________
[gw95] linux -- Python 3.6.12 $PREFIX/bin/python

self = <numpy.core.tests.test_numeric.TestClip object at 0x0000aaab08b4b7c0>

    @given(data=st.data(), shape=hynp.array_shapes())
>   def test_clip_property(self, data, shape):
E   hypothesis.errors.FailedHealthCheck: Data generation is extremely slow: Only produced 9 valid examples in 1.03 seconds (0 invalid ones and 2 exceeded maximum size). Try decreasing size of the data you're generating (with e.g.max_size or max_leaves parameters).
E   See https://hypothesis.readthedocs.io/en/latest/healthchecks.html for more information about this. If you want to disable just this health check, add HealthCheck.too_slow to the suppress_health_check settings for this test.

../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/site-packages/numpy/core/tests/test_numeric.py:2113: FailedHealthCheck
---------------------------------- Hypothesis ----------------------------------
You can add @seed(54826825627429603864147793465765942345) to this test or run pytest with --hypothesis-seed=54826825627429603864147793465765942345 to reproduce this failure.
=============================== warnings summary ===============================
```

</details>

<details>
<summary>win + blis: 11-13 errors</summary>

```
=========================== short test summary info ===========================
FAILED core/tests/test_multiarray.py::TestMatmul::test_matmul_object - Assert...
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestPinv::test_nonsq_cases - AssertionErr...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_sq_cases - Ass...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_nonsq_cases - ...
FAILED linalg/tests/test_linalg.py::TestPinvHermitian::test_generalized_herm_cases
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 13 failed, 10489 passed, 492 skipped, 19 xfailed, 1 xpassed, 230 warnings in 550.47s (0:09:10) =
```

</details>

h-vetinari commented:
For those who might have subscribed to this PR, I'm continuing this for 1.20 in #227

h-vetinari commented:
No more 1.19 patch releases expected, time to close this PR

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict.
Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
### Revival (new PyPy builds and BLAS updates): all green except PPC

Due to rebuilding 1.19 for pypy3.8/3.9, much less several relevant BLAS (& infrastructure) changes, I thought I'd revive this PR for one last update.

From 9 failures out of 92 runs, we're now at _**12 failures (PPC-only) out of 108 runs**_.

### Notable

- Added `accelerate` BLAS flavour on osx
- Testing against PyPy 3.8 and 3.9 added everywhere but for osx-arm
- Big bumps for openblas, blis & MKL
- In the meantime, the previous blis errors have been tracked down and fixed
- The broken pipe errors on windows should be gone thanks to the pagefile work
- Fixed the error reporting of the test suite (used to be broken due to missing `sys.exit` wrapper)
- [Switched](https://github.com/conda-forge/numpy-feedstock/commit/f59124033d3515efa9d59e349b1972d8ed71bd30) to running the full test suite; emulation keeps running only `label='fast'` tests.

### Details

| variant | [before](https://github.com/conda-forge/numpy-feedstock/pull/196#issuecomment-767813877) | after |
|--------|-------|----|
| linux + aarch + pypy3.6 + openblas  | one test failure (probably flaky as it's "only" from hypothesis) | gone 🥳 |
| linux + ppc + pypy + netlib | segfault | gone 🥳 |
| linux + ppc | fine otherwise (on travis) | test failures due to emulation problems (on azure) |
| win + blis | 11-12 failures across all python versions | fixed 🥳  |
| win | 3 failures due to `The process tried to write to a nonexistent pipe.` | gone 🥳 |
| win + pypy | ... | two spurious failures in `test_may_share_memory_easy_fuzz`, resolved by restart |

--------------------

| lib | [before](https://github.com/conda-forge/numpy-feedstock/pull/196#issuecomment-767813877) | after | updated<br/>version | updated<br/>build |
|--------|-------|----|-|-|
| `numpy` | `1.19.5` | `1.19.5` | | |
| `libblas` | `3.9.0-7` | `3.9.0-14` | | X |
| `blis` | `0.8.0-1` | `0.9.0-0` | X | |
| `openblas` | `0.3.12-pthreads-1` | `0.3.20-pthreads-0` | X | |
| `mkl` | `2020.4-304` | `2022.0.1-803` | X | |
| `netlib` | `3.9.0-3` | `3.9.0-5` | | X |
| `pypy` | `7.3.3-3` | `7.3.9-1` (pypy38/39)<br>`7.3.7-3` (pypy37) | X | |
| [`qemu-user-static`](https://github.com/multiarch/qemu-user-static) | ? | `6.1.0-8` | | |

---------------

| variant | accelerate | blis | mkl | netlib | openblas | sum* | 
|--------|------------|----|-----|-------|-----------|-------|
| linux / x86 | ➖ | ✔️ | ✔️ | ✔️ | ✔️ | - | 
| linux / aarch | ➖ | ➖ | ➖ | ✔️ | ✔️ | - | 
| linux / ppc64le | ➖ | ➖ | ➖ | ✖️ | ✖️ | 12F  |
| osx / arm | ✔️ | ➖ | ➖ | ✔️ | ✔️ | - |
| osx / x86 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | - |
| win / x86 | ➖ | ✔️ | ✔️ | ✔️ | ✔️ | - |
| sum* | - | - | - | 6F | 6F | 12F |

`*` sum of Failures (out of a total of 108 CI combinations being tested)

Build logs:
[Azure](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=491288&view=results)
